tests/paramgrill.c

```diff
@@ -632,131 +1391,1112 @@ int benchFiles(const char** fileNamesTable, int nbFiles)
+        benchedSize = BMK_findMaxMem(totalSizeToLoad * 3) / 3;
+
+        origBuff = malloc(benchedSize);
+        if(!origBuff || !fileSizes) {
+            DISPLAY("Not enough memory for stuff\n");
+            ret = 1;
+            goto _cleanUp;
+        }
+        ec = BMK_loadFiles(origBuff, benchedSize, fileSizes, fileNamesTable, (U32)nbFiles);
+        if(ec) {
+            DISPLAY("Error Loading Files");
+            ret = ec;
+            goto _cleanUp;
+        }
+        benchedSize = 0;
+        for(i = 0; i < nbFiles; i++) {
+            benchedSize += fileSizes[i];
+        }
+        origBuff = realloc(origBuff, benchedSize);
+    }
+
+    allMT = memoTableInitAll(paramTarget, target, varArray, varLen, benchedSize);
+    if(!allMT) {
+        ret = 2;
+        goto _cleanUp;
+    }
+ 
+    if(cLevel) {
+        BMK_result_t candidate;
+        const size_t blockSize = g_blockSize ? g_blockSize : benchedSize;
+        ZSTD_CCtx* const ctx = ZSTD_createCCtx();
+        ZSTD_DCtx* const dctx = ZSTD_createDCtx();
+        ZSTD_compressionParameters const CParams = ZSTD_getCParams(cLevel, blockSize, dictBufferSize);
+        if(BMK_benchParam(&candidate, origBuff, benchedSize, fileSizes, (U32)nbFiles, ctx, dctx, CParams)) {
+            ZSTD_freeCCtx(ctx);
+            ZSTD_freeDCtx(dctx);
+            ret = 3;
+            goto _cleanUp;
+        }
+
+        target.cSpeed = (U32)candidate.cSpeed; //Maybe have a small bit of slack here, like x.99? 
+        BMK_printWinner(stdout, cLevel, candidate, CParams, benchedSize);
+
+        ZSTD_freeCCtx(ctx);
+        ZSTD_freeDCtx(dctx);
+    }
+
+    g_targetConstraints = target;
 
     /* bench */
     DISPLAY("\r%79s\r", "");
-    DISPLAY("optimizing for %s - limit speed %u MB/s \n", inFileName, targetSpeed);
-    targetSpeed *= 1000000;
+    if(nbFiles == 1) {
+        DISPLAY("optimizing for %s", fileNamesTable[0]);
+    } else {
+        DISPLAY("optimizing for %lu Files", (unsigned long)nbFiles);
+    }
+    if(target.cSpeed != 0) { DISPLAY(" - limit compression speed %u MB/s", target.cSpeed / 1000000); }
+    if(target.dSpeed != 0) { DISPLAY(" - limit decompression speed %u MB/s", target.dSpeed / 1000000); }
+    if(target.cMem != (U32)-1) { DISPLAY(" - limit memory %u MB", target.cMem / 1000000); }
+    DISPLAY("\n");
+    findClockGranularity();
+
     {   ZSTD_CCtx* const ctx = ZSTD_createCCtx();
         ZSTD_DCtx* const dctx = ZSTD_createDCtx();
         winnerInfo_t winner;
-        BMK_result_t candidate;
+        U32 varNew[NUM_PARAMS];
         const size_t blockSize = g_blockSize ? g_blockSize : benchedSize;
+        U32 const maxNbBlocks = (U32) ((benchedSize + (blockSize-1)) / blockSize) + 1;
+        const size_t maxCompressedSize = ZSTD_compressBound(benchedSize) + (maxNbBlocks * 1024);
+        void* compressedBuffer = malloc(maxCompressedSize);
 
         /* init */
         if (ctx==NULL) { DISPLAY("\n ZSTD_createCCtx error \n"); free(origBuff); return 14;}
```

Comment:
what about `dctx` ?
if it's allocated, that's a memory leak.

Consider nesting, or using `goto`, for proper cleaning before leaving the function.